### PR TITLE
Mapping argument for renaming ingested files

### PR DESCRIPTION
**Pull request by ktsitsi**
This PR:

- Enables the user to provide a name mapping between the source files to be ingested and the desired final ingested name. This feature can be used also for de-identification cases. The user should be able to provide a list of output files given a equal length source file list where each of the source files will be renamed to the corresponding output item.
- Adds an extra layer of validation useful not only for the programmatic API but also for the 1-click ingestion feature.

### Changes

- Generalization and new semantics for the provided `output` and `input` ingestor argument. If the user want to point to a folder for storing the ingested data, then the path provided should have a trailing `/`. If so the default behaviour is to ingest the source files with the same name in this directory.
- If the `output` arg does not contain a trailing `/` then it is considered as a full target path for the ingested source. In this case the ingested source file should be one at a time, otherwise errors out with a `ValueError`.

### CODE REVIEW

File: src/tiledb/cloud/bioimg/helpers.py [hunk 1]
```diff
@@ -34,3 +34,89 @@ def serialize_filter(filter):
         return filter_dict
     else:
         raise TypeError
+
+
+def is_folder(path: str) -> bool:
+    return path.endswith("/")
+
+
+def validate_io_paths(source, output):
```
Comment (ihnorton) [inline src/tiledb/cloud/bioimg/helpers.py:43]: This is very hard to follow. Please break it down into constituent pieces.

Also, I'm not sure why we are supporting multiple output paths. Unless there is a driving use-case, we may need to take a step back and refactor this API to take separate `source: Union[str, List[str]]` and `dest: str` arguments -- in other words, only allow a single output destination (as a separate argument). I think that will simplify this issue signficantly.

Comment (ktsitsi) [reply src/tiledb/cloud/bioimg/helpers.py:43]: The behaviour till now was indeed to support a single output. However by supporting multiple outputs given multiple inputs we are able to support the Quest's use-case, where the renaming of the files comes as a direct consequence. I get that the code gets complicated so I will try to simplify it further however this will probably need further restrictions on what the user can give as arguments.

Comment (Shelnutt2) [reply src/tiledb/cloud/bioimg/helpers.py]: Lets move this to a design doc. We are well overdue on delivering this, I'd like to get a design agreed upon and implementation shipped, validated and accepted before end of week. Please create a document today that covers the use cases and the proposal for what we are to support.

**Comment by sgillies** [comment]
@ktsitsi I don't see the mapping argument in the changes yet, am I overlooking it?

I've got a suggestion for how to think about this feature: [tar transforms](https://www.gnu.org/software/tar/manual/html_section/transform.html). Tar is time-tested and its design has been approved (or, at least, tolerated) by lots of developers. Path stripping is tar's simple form, regex is the advanced form.

This might be generally useful for other kinds of ingest, too.

File: src/tiledb/cloud/bioimg/helpers.py (hunk 1 above)
Comment (ktsitsi) [reply src/tiledb/cloud/bioimg/helpers.py]: https://docs.google.com/document/d/1QDIF0T1mWsaOVswsARy5NKGHXc-Vn98083Sv5dWt6MM/edit?usp=sharing

**Comment by ktsitsi** [comment]
> @ktsitsi I don't see the mapping argument in the changes yet, am I overlooking it?
> 
> I've got a suggestion for how to think about this feature: [tar file name transforms](https://www.gnu.org/software/tar/manual/html_section/transform.html). Tar is time-tested and its design has been approved (or, at least, tolerated) by lots of developers. Path stripping is tar's simple form, regex is the advanced form. Would those cover the use cases?
> 
> This might be generally useful for other kinds of ingest, too.

@sgillies thank you for taking a look. I have updated the PR's description to reflect the latest changes. A mapping argument would be an overkill since the user can utilise list of files as outputs given list of input paths. This will allow them to rename the files being ingested. The API is being used by clients so we any change we do needs to keep that in mind. In case we would like to change the overall thinking we will need to take into account that we need to abstract the implementation from the user. I am open though to explore alternatives as long as the former is satisfied.

**Review by ihnorton — CHANGES_REQUESTED** [review]
Per discussion, we need to add a recursion depth limit (and set it to 1 for cloud ingestion).

File: src/tiledb/cloud/bioimg/helpers.py [hunk 2]
```diff
@@ -34,3 +35,29 @@ def serialize_filter(filter):
...
+
+    if len(output) == 1 and is_folder(output[0]):
+        return True
+
+    if all(is_folder(path) for path in output):
+        if all(is_folder(path) for path in source):
+            if len(output) == len(source):
+                return True
+
+    if all(not is_folder(path) for path in output):
```
Comment (ihnorton) [inline src/tiledb/cloud/bioimg/helpers.py:59]: This is the negation of condition on `L54`, use `else:`

File: src/tiledb/cloud/bioimg/ingestion.py [hunk 3]
```diff
@@ -48,19 +52,20 @@ def ingest(
     :param namespace: The namespace where the DAG will run
     :param verbose: verbose logging, defaults to False
     :param output_ext: extension for the output images in tiledb
+    :param traverse: If true then traverse the src paths for exploring images in
```
Comment (ihnorton) [inline src/tiledb/cloud/bioimg/ingestion.py:55]:
```suggestion
    :param traverse: If true then traverse the src paths recursively to find images to convert in
```

File: src/tiledb/cloud/bioimg/helpers.py (hunk 2 above)
Comment (ktsitsi) [reply src/tiledb/cloud/bioimg/helpers.py:59]: The L54 checks if all paths are `folders` where L59 checks if all paths are `files`.  For example a case that hasn't the same result if `else` is used is the following:
For an input like `test3 = [False, True, True]`
```
if all(i for i in test3):
    print('A')
if all(not i for i in test3):
    print('B')
```
prints `nothing` whereas:
```
if all(i for i in test3):
    print('A')
else:
    print('B')
``` 
prints `B`. The test pass even with `else` cause the rest of the code handles it, but it was intended.

**Comment by ktsitsi** [comment]
Closing it in favor of #509 .
